### PR TITLE
Expose oc build timeout in preference

### LINKED ISSUE #3110: Expose OcBuildTimeout to overwrite the default value
/kind feature

## Which functionality do you think we should add?
Fix the reason behind the flake https://github.com/openshift/odo/issues/2877. Actually one user found the same issue on his setup (Odo version : v1.1.3) and shared the detail analysis

User says:
```
I can confirm that the problem is the build timeout in odo.

I have modified the odo client, cloning the github repo, modified the file pkg/occlient/occlient.go and
 increase the value from the variable OcBuildTimeout in line 99 to 15 (default was 5), and the build
 finished successfully, injecting the resulting image name into the deploymentconfig.

diff --git a/pkg/occlient/occlient.go b/pkg/occlient/occlient.go
index 6e59dd646..637016acc 100644
--- a/pkg/occlient/occlient.go
+++ b/pkg/occlient/occlient.go
@@ -95,8 +95,8 @@ type CreateArgs struct {

 const (
        failedEventCount                = 5
-       OcUpdateTimeout                 = 5 * time.Minute
-       OcBuildTimeout                  = 5 * time.Minute
+       OcUpdateTimeout                 = 15 * time.Minute
+       OcBuildTimeout                  = 15 * time.Minute
        OpenShiftNameSpace              = "openshift"
        waitForComponentDeletionTimeout = 120 * time.Second

I believe the timeout in the default value is reached due to the amount of maven packages the build
 needs to download, and since our connection is not the fastest, the build reach this timeout and 
exits.

Perhaps you can add an option in the future where the user can define the timeout value for the build
 in slow or low processing environments.
```
## Why is this needed?
Now its clear that the issue not with the current odo backend code rather its an environment specific issue (I mean slow env or slow network), however we should expose a key that would overwrite the default build time out in case of slower environment. Also revisit the default ```OcBuildTimeout``` value.

### PR DESCRIPTION
**Pull request by amitkrout**
**What type of PR is this?**
/kind feature

**What does does this PR do / why we need it**:

**Which issue(s) this PR fixes**:

Fixes part of #3110 as per @mik-dass https://github.com/openshift/odo/issues/2877#issuecomment-627821138 

**How to test changes / Special notes to the reviewer**:
CI should not fail with build timeout error

### CODE REVIEW

**Comment by amitkrout** [comment]
prow issue
/test v4.3-integration-e2e

**Comment by amitkrout** [comment]
The code changes has been done, so i am opening it for review though i have to write integration test.

File: pkg/preference/preference.go [hunk 1]
```diff
@@ -86,6 +92,9 @@ var TimeoutSettingDescription = fmt.Sprintf("Timeout (in seconds) for OpenShift
 // PushTimeoutSettingDescription adds a description for PushTimeout
 var PushTimeoutSettingDescription = fmt.Sprintf("PushTimeout (in seconds) for waiting for a Pod to come up (Default: %d)", DefaultPushTimeout)
 
+// BuildTimeoutSettingDescription adds a description for BuildTimeout
+var BuildTimeoutSettingDescription = fmt.Sprintf("BuildTimeout (in seconds) for waiting for a build to complete (Default: %d)", DefaultBuildTimeout)
```
Comment (kadel) [inline pkg/preference/preference.go:96]: BuildTimeout is valid only for git components, The help message should mention that.

File: pkg/occlient/occlient.go [hunk 2]
```diff
@@ -1746,7 +1745,14 @@ func (c *Client) WaitForBuildToFinish(buildName string, stdout io.Writer) error
 	// following indicates if we have already setup the following logic
 	following := false
 	klog.V(4).Infof("Waiting for %s  build to finish", buildName)
-
+	// Try to grab the preference in order to set a timeout.. but if not, we’ll use the default.
+	buildTimeout := preference.DefaultBuildTimeout * time.Second
+	cfg, configReadErr := preference.New()
+	if configReadErr != nil {
+		klog.V(4).Info(errors.Wrap(configReadErr, "unable to read config file"))
```
Comment (girishramnani) [inline pkg/occlient/occlient.go:1752]: preference file

File: pkg/occlient/occlient.go [hunk 3]
```diff
@@ -2021,8 +2027,17 @@ func (c *Client) FollowBuildLog(buildName string, stdout io.Writer) error {
 		NoWait: false,
 	}
 
+	// Try to grab the preference in order to set a timeout.. but if not, we’ll use the default.
+	buildTimeout := preference.DefaultBuildTimeout * time.Second
+	cfg, configReadErr := preference.New()
```
Comment (girishramnani) [inline pkg/occlient/occlient.go:2032]: please initialise the preference once as part of the struct or as a singleton.

Comment (girishramnani) [reply pkg/occlient/occlient.go:2032]: I can already 2 more places where preference has been read and initialised

File: pkg/preference/preference.go [hunk 4]
```diff
@@ -420,6 +443,15 @@ func (c *PreferenceInfo) GetTimeout() int {
 	return *c.OdoSettings.Timeout
 }
 
+// GetBuildTimeout gets the value set by BuildTimeout
+func (c *PreferenceInfo) GetBuildTimeout() int {
+	// default timeout value is 1
```
Comment (girishramnani) [inline pkg/preference/preference.go:448]: default timeout value is 1? seems like `300` is the correct value

**Comment by girishramnani** [comment]
Any update on this PR @amitkrout

**Comment by girishramnani** [comment]
@amitkrout Any update on this PR?

**Comment by amitkrout** [comment]
I have addressed all the review comments. Ping @kadel @girishramnani

File: pkg/component/component.go [hunk 5]
```diff
@@ -712,6 +713,15 @@ func PushLocal(client *occlient.Client, componentName string, applicationName st
 // 'show' will determine whether or not the log will be shown to the user (while building)
 func Build(client *occlient.Client, componentName string, applicationName string, wait bool, stdout io.Writer, show bool) error {
 
+	// Try to grab the preference in order to set a timeout.. but if not, we’ll use the default.
+	buildTimeout := preference.DefaultBuildTimeout * time.Second
+	cfg, configReadErr := preference.New()
+	if configReadErr != nil {
+		klog.V(4).Info(errors.Wrap(configReadErr, "unable to read config file"))
+	} else {
+		buildTimeout = time.Duration(cfg.GetTimeout()) * time.Second
```
Comment (mik-dass) [inline pkg/component/component.go:722]: `cfg.GetBuildTimeout()`

Comment (amitkrout) [reply pkg/component/component.go:722]: yes

File: pkg/occlient/occlient.go [hunk 6]
```diff
@@ -1895,7 +1894,7 @@ func (c *Client) WaitAndGetPod(selector string, desiredPhase corev1.PodPhase, wa
 	if configReadErr != nil {
 		klog.V(4).Info(errors.Wrap(configReadErr, "unable to read config file"))
 	} else {
-		pushTimeout = time.Duration(cfg.GetPushTimeout()) * time.Second
+		pushTimeout = time.Duration(cfg.GetTimeout()) * time.Second
```
Comment (mik-dass) [inline pkg/occlient/occlient.go:1897]: Don't change this to GetTimeout()

Comment (amitkrout) [reply pkg/occlient/occlient.go]: In my previous commit i missed it. Thanks for the pointer

Done

**Comment by girishramnani** [comment]
@amitkrout please add this field into machine readable out for preference

**Comment by amitkrout** [comment]
> @amitkrout please add this field into machine readable out for preference

Done

**Comment by amitkrout** [comment]
```
 gosec -severity medium -confidence medium -exclude G304,G204 -quiet  ./...
golangci-lint run ./... --timeout 5m
make[1]: golangci-lint: Command not found
make[1]: *** [golint] Error 127
make[1]: Leaving directory `/go/src/github.com/openshift/odo'
make: *** [openshiftci-presubmit-unittests] Error 2
--- 
```
/test unit

**Comment by mik-dass** [comment]
The unit is failing
```
.036s [no tests to run]
==================
WARNING: DATA RACE
Write at 0x00c00021fe10 by goroutine 28:
  runtime.closechan()
      /usr/local/go/src/runtime/chan.go:335 +0x0
  github.com/openshift/odo/vendor/k8s.io/apimachinery/pkg/watch.(*FakeWatcher).Stop()
      /home/mrinaldas/go/src/github.com/openshift/odo/vendor/k8s.io/apimachinery/pkg/watch/watch.go:115 +0x116
  github.com/openshift/odo/pkg/occlient.(*Client).WaitForBuildToFinish()
      /home/mrinaldas/go/src/github.com/openshift/odo/pkg/occlient/occlient.go:1789 +0xabc
  github.com/openshift/odo/pkg/occlient.TestWaitForBuildToFinish.func1()
      /home/mrinaldas/go/src/github.com/openshift/odo/pkg/occlient/occlient_test.go:2186 +0x442
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:991 +0x1eb

Previous read at 0x00c00021fe10 by goroutine 29:
  runtime.chansend()
      /usr/local/go/src/runtime/chan.go:142 +0x0
  github.com/openshift/odo/vendor/k8s.io/apimachinery/pkg/watch.(*FakeWatcher).Modify()
      /home/mrinaldas/go/src/github.com/openshift/odo/vendor/k8s.io/apimachinery/pkg/watch/watch.go:145 +0x1ed
  github.com/openshift/odo/pkg/occlient.TestWaitForBuildToFinish.func1.1()
      /home/mrinaldas/go/src/github.com/openshift/odo/pkg/occlient/occlient_test.go:2179 +0x2f

Goroutine 28 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1042 +0x660
  github.com/openshift/odo/pkg/occlient.TestWaitForBuildToFinish()
      /home/mrinaldas/go/src/github.com/openshift/odo/pkg/occlient/occlient_test.go:2173 +0x302
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:991 +0x1eb

Goroutine 29 (running) created at:
  github.com/openshift/odo/pkg/occlient.TestWaitForBuildToFinish.func1()
      /home/mrinaldas/go/src/github.com/openshift/odo/pkg/occlient/occlient_test.go:2178 +0x14d
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:991 +0x1eb
==================
panic: send on closed channel
```

**Comment by mik-dass** [comment]
The unit is failing
```
.036s [no tests to run]
==================
WARNING: DATA RACE
Write at 0x00c00021fe10 by goroutine 28:
  runtime.closechan()
      /usr/local/go/src/runtime/chan.go:335 +0x0
  github.com/openshift/odo/vendor/k8s.io/apimachinery/pkg/watch.(*FakeWatcher).Stop()
      /home/mrinaldas/go/src/github.com/openshift/odo/vendor/k8s.io/apimachinery/pkg/watch/watch.go:115 +0x116
  github.com/openshift/odo/pkg/occlient.(*Client).WaitForBuildToFinish()
      /home/mrinaldas/go/src/github.com/openshift/odo/pkg/occlient/occlient.go:1789 +0xabc
  github.com/openshift/odo/pkg/occlient.TestWaitForBuildToFinish.func1()
      /home/mrinaldas/go/src/github.com/openshift/odo/pkg/occlient/occlient_test.go:2186 +0x442
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:991 +0x1eb

Previous read at 0x00c00021fe10 by goroutine 29:
  runtime.chansend()
      /usr/local/go/src/runtime/chan.go:142 +0x0
  github.com/openshift/odo/vendor/k8s.io/apimachinery/pkg/watch.(*FakeWatcher).Modify()
      /home/mrinaldas/go/src/github.com/openshift/odo/vendor/k8s.io/apimachinery/pkg/watch/watch.go:145 +0x1ed
  github.com/openshift/odo/pkg/occlient.TestWaitForBuildToFinish.func1.1()
      /home/mrinaldas/go/src/github.com/openshift/odo/pkg/occlient/occlient_test.go:2179 +0x2f

Goroutine 28 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1042 +0x660
  github.com/openshift/odo/pkg/occlient.TestWaitForBuildToFinish()
      /home/mrinaldas/go/src/github.com/openshift/odo/pkg/occlient/occlient_test.go:2173 +0x302
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:991 +0x1eb

Goroutine 29 (running) created at:
  github.com/openshift/odo/pkg/occlient.TestWaitForBuildToFinish.func1()
      /home/mrinaldas/go/src/github.com/openshift/odo/pkg/occlient/occlient_test.go:2178 +0x14d
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:991 +0x1eb
==================
panic: send on closed channel
```
/hold

File: pkg/occlient/occlient_test.go [hunk 7]
```diff
@@ -2183,7 +2183,7 @@ func TestWaitForBuildToFinish(t *testing.T) {
 				return true, fkWatch, nil
 			})
 
-			err := fkclient.WaitForBuildToFinish(tt.buildName, os.Stdout)
+			err := fkclient.WaitForBuildToFinish(tt.buildName, os.Stdout, 5)
```
Comment (mik-dass) [inline pkg/occlient/occlient_test.go:2186]: `time.Second*5`

File: pkg/preference/preference.go [hunk 8]
```diff
@@ -334,6 +347,16 @@ func (c *PreferenceInfo) SetConfiguration(parameter string, value string) error
 			}
 			c.OdoSettings.Timeout = &typedval
 
+		case "buildtimeout":
```
Comment (mik-dass) [inline pkg/preference/preference.go:350]: please update the unit test for setConfiguration to test buildTimeout too

Comment (amitkrout) [reply pkg/preference/preference.go:350]: 💯 
Done

File: pkg/preference/preference_test.go [hunk 9]
```diff
@@ -346,7 +346,37 @@ func TestSetConfiguration(t *testing.T) {
 			wantErr:        true,
 		},
 		{
-			name:           fmt.Sprintf("Case 12: %s set to 300 with mixed case in parameter name", TimeoutSetting),
+			name:           fmt.Sprintf("Case 11: %s set to 50 with mixed case in parameter name", TimeoutSetting),
```
Comment (mik-dass) [inline pkg/preference/preference_test.go:349]: case 12

Comment (amitkrout) [reply pkg/preference/preference_test.go]: hawk eye

**Pull request merged**